app/src/main/java/org/thoughtcrime/securesms/util/FlowUtils.kt

```diff
@@ -0,0 +1,42 @@
+package org.thoughtcrime.securesms.util
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.flatMapConcat
+
+/**
+ * Buffers items from the flow and emits them in batches. The batch will have size [maxItems] and
+ * time [timeoutMillis] limit.
+ */
+fun <T> Flow<T>.timedBuffer(timeoutMillis: Long, maxItems: Int): Flow<List<T>> {
+    return channelFlow {
+        val buffer = mutableListOf<T>()
+        var bufferBeganAt = -1L
+
+        collectLatest { value ->
+            if (bufferBeganAt < 0) {
```

Comment:
```suggestion
            if (buffer.isEmpty()) {
```

semantically, maybe `if (buffer.isEmpty())` is more meaningful, and then you don't need to reset `bufferBeganAt = -1L` at the end etiher.